service/history/workflow/mutable_state_impl.go

```diff
@@ -4955,31 +4972,80 @@ func (ms *MutableStateImpl) updateActivityInfoForRetries(
 	nextAttempt int32,
 	activityFailure *failurepb.Failure,
 ) {
-	ms.updateActivity(ai, func(info *persistencespb.ActivityInfo, impl *MutableStateImpl) *persistencespb.ActivityInfo {
-		ai = UpdateActivityInfoForRetries(
-			ai,
-			ms.GetCurrentVersion(),
-			nextAttempt,
-			ms.truncateRetryableActivityFailure(activityFailure),
-			timestamppb.New(nextScheduledTime),
-		)
-		return ai
+	_ = ms.UpdateActivityWithCallback(ai, func(info *persistencespb.ActivityInfo, mutableState MutableState) {
+		mutableStateImpl, ok := mutableState.(*MutableStateImpl)
+		if ok {
+			ai = UpdateActivityInfoForRetries(
+				ai,
+				mutableStateImpl.GetCurrentVersion(),
+				nextAttempt,
+				mutableStateImpl.truncateRetryableActivityFailure(activityFailure),
+				timestamppb.New(nextScheduledTime),
+			)
+		}
 	})
 }
 
-func (ms *MutableStateImpl) updateActivity(ai *persistencespb.ActivityInfo, updateCallback func(*persistencespb.ActivityInfo, *MutableStateImpl) *persistencespb.ActivityInfo) {
-	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity
-	// info and if prev found it points to the same activity info as ai, so updating ai will cause
-	// size of prev change.
+func (ms *MutableStateImpl) UpdateActivityWithCallback(
+	ai *persistencespb.ActivityInfo,
+	updateCallback UpdateActivityCallback,
+) error {
+	// incoming activity info can be a pointer to the activity info in pendingActivityInfoIDs
+	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity info
+	// if prev found it can point to the same activity info as incoming activity info,
+	prevPause := ai.Paused
 	var originalSize int
 	if prev, ok := ms.pendingActivityInfoIDs[ai.ScheduledEventId]; ok {
 		originalSize = prev.Size()
+		prevPause = prev.Paused
 	}
+
 	updateCallback(ai, ms)
 
+	if prevPause != ai.Paused {
+		err := ms.UpdatePauseInfoSearchAttribute()
+		if err != nil {
+			return err
+		}
+	}
+
 	ms.approximateSize += ai.Size() - originalSize
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
 	ms.syncActivityTasks[ai.ScheduledEventId] = struct{}{}
+
+	return nil
+}
+
+func (ms *MutableStateImpl) UpdatePauseInfoSearchAttribute() error {
+	pausedInfoMap := make(map[string]struct{})
+
+	for _, ai := range ms.GetPendingActivityInfos() {
+		if !ai.Paused {
+			continue
+		}
+		pausedInfoMap[ai.ActivityType.Name] = struct{}{}
+	}
+	pausedInfo := make([]string, 0, len(pausedInfoMap))
+	for activityType := range pausedInfoMap {
+		pausedInfo = append(pausedInfo, fmt.Sprintf("entityType.activity:%s", activityType))
```

Comment:
nit: I still think using `:` as the sole namespace delimiter is the right way to go here.

Reply:
will discuss separately.

Reply:
discussed.
agree on
```
“property:activityType=MYBESTACTIVITYTYPE”

“property:workflowType=MYWORKFLOWID”
```